Add Arcee (AFM) model support to vLLM

alyosha-swamy opened the pull request:
# [New Model] Support Arcee (Arcee Foundational Models)

## 1. Purpose (Why this PR?)
Add inference support for Arcee Foundational Model (AFM) so that users can serve it with vLLM in both Python and API-server workflows. AFM uses a unique ReLU² activation in its MLP layers, differentiating it from standard Llama-based models.

## 2. Model details

| Field | Value / Reference |
|-------|------------------|
| Source repo / HF id | huggingface.co/arcee-ai/AFM-4.5B-Base |
| Architecture | Llama-style decoder-only transformer with ReLU² MLP activation |
| Context length | 64k tokens |
| Hidden size / #layers | 4096 / 32 |
| License | CC BY-NC 4.0 |
| Special quirks | Uses ReLU² (squared ReLU) activation instead of SiLU in MLP layers |

## 3. Implementation overview
* Added `ArceeForCausalLM` class in `vllm/model_executor/models/arcee.py` with custom `ArceeMLP` using ReLU² activation
* Registered model in `_TEXT_GENERATION_MODELS` in `vllm/model_executor/models/registry.py`
* Updated `docs/models/supported_models.md` with Arcee entry in text generation table
* Reused `LlamaAttention` from existing Llama implementation for attention layers
* Implemented proper LoRA and Pipeline Parallelism support

## 4. Performance / sanity check
```bash
$ python -m vllm.entrypoints.openai.api_server --model arcee-ai/AFM-4.5B-Base --trust-remote-code
$ curl http://localhost:8000/v1/completions -H "Content-Type: application/json" -d '{
    "model": "arcee-ai/AFM-4.5B-Base",
    "prompt": "The future of artificial intelligence is",
    "max_tokens": 50
}'
```

**Expected:** Coherent completion about life's meaning

**Observed:** " a question that has been asked throughout the history of mankind. The search for an answer to this question has inspired countless works of art, literature, and philosophy. Whether we consider the existentialist ideas of Albert Camus or the religious perspectives of spiritual leaders"

## 5. Test plan ✔️

| Test | Command | Expected |
|------|---------|----------|
| Unit | `pytest tests/models/test_arcee.py` | All tests pass |
| Model Loading | `python -c "from vllm import LLM; llm = LLM('arcee-ai/AFM-4.5B-Base')"` | Model loads without errors |
| Integration | `vllm serve arcee-ai/AFM-4.5B-Base --trust-remote-code` | Server starts, responds to requests |
| Generation | `curl localhost:8000/v1/completions` | 200 OK + valid completions |

## 6. Documentation
* Added row to `docs/models/supported_models.md` under Text Generation models
* Model listed as `ArceeForCausalLM` with example model `arcee-ai/AFM-4.5B-Base`
* Marked as supporting LoRA (✅), Pipeline Parallel (✅), and V1 (✅)

## Checklist
- [x] I ran `pre-commit run --all-files` (ruff formatting)
- [x] All CI tests pass locally (`pytest -q`)
- [x] The PR description follows vLLM's "Essential Elements" template
- [x] No breaking changes for existing model classes

## Notes for reviewers
The key architectural difference from standard Llama models is the MLP activation function. Arcee uses ReLU² (squared ReLU) instead of SiLU:
- `ArceeMLP` implements: `x = torch.pow(torch.relu(x), 2)`
- No gating mechanism (no `gate_proj`), only `up_proj` and `down_proj`
- All other components (attention, layer norm, etc.) reuse existing Llama implementations

The model has been tested with an internal HF repo during development, but the official model is `arcee-ai/AFM-4.5B-Base`.

## Test result

| seq | Prompt | vLLM Output |
|-----|--------|-------------|
| 0 | "The meaning of life is" | " a question that has been asked throughout the history of mankind. The search for an answer to this question has inspired countless works of art, literature, and philosophy. Whether we consider the existentialist ideas of Albert Camus or the religious perspectives of spiritual leaders" |
| 1 | "Climate change is primarily caused by" | " human activity, specifically the emission of greenhouse gases such as carbon dioxide (CO2) and methane (CH4). It leads to changes in average temperatures and weather patterns, impacting both nature and human society." |
| 2 | "Machine learning algorithms work by" | " training a predictive model using labeled training data: the model detects patterns in the training data and learns from it. That model is then tested using a test set, which it must predict to achieve a good accuracy rate." |

All outputs are coherent and contextually appropriate.

alyosha-swamy committed:
Add Arcee (AFM) model support to vLLM

Signed-off-by: alyosha-swamy <raghav@arcee.ai>

alyosha-swamy committed:
Update arcee.py with code review improvements and syntax fix

Signed-off-by: alyosha-swamy <raghav@arcee.ai>

alyosha-swamy committed:
Fix linting issues: line length E501 errors

Signed-off-by: alyosha-swamy <raghav@arcee.ai>

alyosha-swamy committed:
Fix linting issues: line length E501 errors

Signed-off-by: alyosha-swamy <raghav@arcee.ai>

jeejeelee commented on vllm/model_executor/models/arcee.py:
```diff
@@ -0,0 +1,473 @@
+        "qkv_proj": ["q_proj", "k_proj", "v_proj"],
+    }
+    # Supported LoRA modules (same as LLaMA, minus gate_proj)
+    supported_lora_modules = [
```
We removed `supported_lora_modules` a long time ago. Please delete it.

jeejeelee commented:
Please  also update this model in https://github.com/vllm-project/vllm/blob/main/tests/models/registry.py

alyosha-swamy committed:
Address review comments: remove supported_lora_modules and update comment

- Remove deprecated supported_lora_modules attribute
- Add ArceeForCausalLM to test registry

Signed-off-by: alyosha-swamy <raghav@arcee.ai>

jeejeelee commented on vllm/model_executor/models/arcee.py:
```diff
@@ -0,0 +1,469 @@
+    packed_modules_mapping = {
+        "qkv_proj": ["q_proj", "k_proj", "v_proj"],
+    }
+    # (No MLP prefix since there's no gate_proj)
```
```suggestion

```

jeejeelee commented on vllm/model_executor/models/arcee.py:
```diff
@@ -0,0 +1,469 @@
+                self.unpadded_vocab_size += lora_config.lora_extra_vocab_size
+
+            # Import DEFAULT_VOCAB_PADDING_SIZE
+            from vllm.model_executor.layers.vocab_parallel_embedding import (
```
Why import here?"

jeejeelee commented:
Can you fix the pre-commit failure?

hmellor commented:
If this is a private model, please either:

- Write your vLLM modelling code and register the model as an out-of-tree model https://docs.vllm.ai/en/latest/contributing/model/registration.html
- Write your Transformers modelling code according to https://docs.vllm.ai/en/latest/models/supported_models.html#transformers and use `--model-impl transformers`

adarshxs commented:
@hmellor @jeejeelee the model is set to be released as open weights on HF soon, this is therefore for day zero support of the model. The architecture is the same as llama with relu^2 act func.

alyosha-swamy committed:
- Inherit from LlamaForCausalLM for most functionality
- Set is_available_online=False in test registry for CI compatibility
